recipes-kernel/linux/5.4/patches/blktap2.patch

```diff
@@ -386,18 +391,28 @@ PATCHES
 +{
 +	struct blktap *tap;
 +	struct blktap_ring *ring;
++	bool loop = true;
 +
 +	tap = blktaps[minor];
 +	if (!tap)
 +		return -ENODEV;
 +
 +	ring = &tap->ring;
 +
-+	for (;;) {
++	if (ring->vma == NULL && ring->n_pending) {
++		printk(KERN_WARNING
++		       "blktap%d ring not mapped and n_pending %d\n",
++		       minor, ring->n_pending);
++		loop = false;
++	}
++
++	while (loop) {
```

Comment:
Not sure about the discussion you had but if you find the ring NULL with pending requests, then why not flush/zero out the pending request instead of setting `loop = false`? This would mean you wouldn't need the NULL check again in the wait_event_interruptible_timeout.

Reply:
And if there is concerns about request coming before ring is finished getting setup, i would think you could just flush the pending requests with an ERETRY.

Reply:
With the change to`wait_event_interruptible_timeout` to check `ring->vma`, yes, the `if` clause is no longer necessary.  The corresponding `wake_up_interruptible(&tap->remove_wait);` may also be needed.  I modified the code to include the `if` when I was debugging out the issue.

I'm not sure about limiting the retries.  If ring->vma is set, then tapdisk can keep servicing requests and we can make progress.  And whenever n_pending changes (wake_up()s added in https://github.com/OpenXT/xenclient-oe/pull/1359/commits/945ca56bc945cc2c4e9b28345f1e4f86daa4fcd8) then the wait event will be triggered.  So now I think it is fine. Maybe the loop could be removed, and `wait_event_interruptible_timeout` -> `wait_event_interruptible`?  Haven't investigated that though.

Reply:
Right though the logical concern here is if you end up in here with a ring that has gone away and pending requests. Theoretically this could be, 1. the blocktap node came up faster that the ring was set up (if that is even possible) or 2. the ring was torn down before all requests were serviced. In both case I think it would be better to flush the pending requests with ERETRY than just moving along. Perhaps that is done elsewhere and this is a non-issue. At a minimum I would keep the if statement along with the debug printk.

Reply:
I haven't investigated, but I think (hope) case 1 on start-up shouldn't be possible.  Case 2 on shutdown was observed with a hang.  Something (tap-ctl?) run by xenmgr was hanging when destroying the tapdev after reading a kernel out of a .vhd.  umount had already returned.

There is `blktap_ring_fail_pending` to clear out the pending requests.  That is called before ring->vma is cleared.  Maybe the device blk-mq queues need to be stopped at that point to prevent adding more requests?  Presumably in the other block device teardown code any pending requests are cancelled, but, again, I haven't investigated.

The `if` and debug printk were helpful when debugging, but isn't particularly useful when things are actually working :).  I'd just drop it.

Reply:
Apologies I wasn't complete in my printk suggestion. I wasn't meaning that it be left as a warning level, but covert it to a pr_debug. Because as you stated, it is helpful during debug.